### PR TITLE
Map label to PR according to nested path

### PR DESCRIPTION
**Pull request by boaz1337**
**NOTE**: this is an idea that I have to support [this usecase](https://github.com/theforeman/prprocessor/pull/118#discussion_r211153786). I would love to hear what you think.

Currently you can set labels according to one of the root directories (`app`, `webpack`, etc). In other words, you cannot label a PR according to a file in `app/assets`.

This commit adds this feature. For example in `repos.yml`:

```yml
theforeman/foreman:
  directory_labels:
    app/assets: ui
    app/controllers: backend
    webpack: ui
```

it will add `ui` label if there were changed files in `./app/assets/` or `./webpack/` and `backend` label if
a file was changed in `./app/controllers`.

//cc : @ekohl  @ohadlevy  @tbrisker

### CODE REVIEW

**Comment by ekohl** [comment]
I'm not a fan of the nested hash. How about just a simple hash where every key is a path?
```yaml
theforeman/foreman:
  directory_labels:
    app/assets: ui
    app/controllers: backend
    webpack: ui
```

Then you can compare it with something like:
```ruby
def is_subpath_of(matcher, path)
  matcher_parts = matcher.split('/')
  path_parts = path.split('/')
  return path_parts.length >= matcher_parts.length && !matcher_parts.zip(path_parts).any? { |first, second| first != second }
end
```

**Comment by tbrisker** [comment]
Am I missing something here? why not just something like `path.starts_with?(matcher)`?

**Comment by ekohl** [comment]
> Am I missing something here? why not just something like path.starts_with?(matcher)?

Is that a function of the pathing library?

```
[22] pry(main)> 'app/assets2'.starts_with?('app/assets')
NoMethodError: undefined method `starts_with?' for "app/assets2":String
Did you mean?  start_with?
from (pry):32:in `__pry__'
```

Some tests that should pass:
```ruby
refute is_subpath_of('app/assets', 'app')
assert is_subpath_of('app/assets', 'app/assets')
assert is_subpath_of('app/assets', 'app/assets/bla')
refute is_subpath_of('app/assets', 'app/assets2')
refute is_subpath_of('app/assets', 'assets/app')
```

**Comment by tbrisker** [comment]
Sorry, i meant `start_with?` - that's a string function, https://ruby-doc.org/core-2.5.1/String.html#method-i-start_with-3F

**Comment by ekohl** [comment]
That would fail with `app/assets2` matching on `app/assets`

**Comment by ripcurld** [comment]
We might go with @tbrisker idea if we do normalization (adding `/` if not giving in the end of `matcher`) before `path.start_with?(matcher)` unless I am missing an edge-case.

@ekohl thanks for your implementation, it does look way more elegant than mine. I will update the PR and add your tests.

Thanks.

**Comment by tbrisker** [comment]
@ekohl - Like @ripcurld suggests, that edge case could be easily solvable by adding a `/` at the end to make sure only the exact dir is matched. TBH, I'm not sure that edge case is one that exists in any of our code that we want these notifications for.

File: github/pull_request.rb [hunk 1]
```diff
@@ -179,9 +179,17 @@ def add_status(state, message, options = {})
     @client.create_status(repo.full_name, options[:sha], state, context: 'prprocessor', description: message, target_url: options[:url])
   end
 
+  def get_labels(filename, mapping)
+    mapping.select { |k, v| filename.start_with?(k) }.map { |k, v| v }
```
Comment (ekohl) [inline github/pull_request.rb]: This should be `filename == k || filename.start_with("#{k}/")`. That way you can also match files. You can then remove the normalization code I think.

You can also use `.values` instead of the map.

Comment (boaz1337) [reply github/pull_request.rb]: Thanks @ekohl 
Just one problem that I foresee - "directory_label" will be very confusing if we set there files names too. Maybe we should mention this in a document or rename it to "path_label"? 

WDYT?

Comment (ekohl) [reply github/pull_request.rb]: :+1: for `path_labels`

File: github/pull_request.rb [hunk 2]
```diff
@@ -179,9 +179,17 @@ def add_status(state, message, options = {})
     @client.create_status(repo.full_name, options[:sha], state, context: 'prprocessor', description: message, target_url: options[:url])
   end
 
-  def set_directory_labels(mapping)
+  def get_labels(filename, mapping)
+    mapping.select { |k, v| filename == k || filename.start_with?((k+"/").squeeze) }.values
```
Comment (ekohl) [inline github/pull_request.rb]: I didn't know squeeze but isn't this dangerous? Looks like it squeezes all duplicate characters but in a directory name you can have them for valid reasons (`lookups` would become `lokups`).

Comment (boaz1337) [reply github/pull_request.rb]: My bad. I thought `squeeze` removes characters from the end of the string.
What I wanted to do is to remove extra `/` in the key. For example, if someone adds "app/views/" to the map then it probably won't add "app/views/" label because the file "app/views/models/index.html" doesn't start with "app/views//".

Maybe using `gsub` can be replaced here?

**Comment by boaz1337** [comment]
@ekohl rebased and updated - can you give feedback again? thanks :smiley:

File: github/pull_request.rb [hunk 3]
```diff
@@ -187,9 +187,17 @@ def add_status(state, message, options = {})
     @client.create_status(repo.full_name, options[:sha], state, context: 'prprocessor', description: message, target_url: options[:url])
   end
 
-  def set_directory_labels(mapping)
+  def get_labels(filename, mapping)
+    mapping.select { |k, v| filename == k || filename.start_with?((k+"/").gsub(/\/+$/, "/")) }.values
```
Comment (ekohl) [inline github/pull_request.rb]: I don't like the double trailing slash cleaning here. Can't we just require the configuration to be correct? We can introduce a test on `config/repos.yaml` that fails if there's a trailing slash in a mapping. That way Travis will fail on a PR that attempts to introduce an invalid config.

File: test/config_test.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+require 'repository'
+
+class TestRepoConfig < Minitest::Test
+  def test_path_labels_keys_does_not_end_with_backslash
+    paths = Repository.all.values
+      .select{ |r| r.path_labels? }
+      .map{ |r| r.path_labels.keys }
+      .flatten
+      .select { |file_path| file_path.end_with?('/') }
+    assert_equal [], paths, "remove backslash from the following paths: #{paths.join(', ')}"
```
Comment (ekohl) [inline test/config_test.rb]: It's a slash (`/`), not a backslash (`\`). I'd also mention it's a trailing slash. Note backslash is also in the test name.

**Comment by boaz1337** [comment]
@ekohl done
